File: sdk/python/receipt.py
"""Reference Phase 1B AVR prototype canonicalization and identifier derivation."""

import hashlib
import json
import re
from datetime import datetime

RECEIPT_DOMAIN = "aichain:avr:0.1.0-draft:"
COMMITMENTS_DOMAIN = "aichain:avr:commitments:0.1.0-draft:"
ATTESTATION_MESSAGE_DOMAIN = "AIChain AVR v0.1.0-draft receipt: "
RECEIPT_FIELDS = {"schema", "schemaVersion", "assuranceLevel", "issuer", "execution", "commitments"}
COMMITMENT_FIELDS = {"configuration", "input", "model", "output", "policy", "provider"}
BYTES32 = re.compile(r"^0x[0-9a-fA-F]{64}$")
ADDRESS = re.compile(r"^0x[0-9a-fA-F]{40}$")


def validate_receipt(receipt: dict) -> None:
    """Validate the fixed Phase 1B draft shape before canonicalization."""
    if not isinstance(receipt, dict):
        raise ValueError("Receipt must be an object")
    payload = {key: value for key, value in receipt.items() if key not in ("expected", "attestation")}
    if set(payload) != RECEIPT_FIELDS:
        raise ValueError("Receipt fields do not match the AVR v0.1.0-draft shape")
    if payload["schema"] != "aichain.avr" or payload["schemaVersion"] != "0.1.0-draft":
        raise ValueError("Unsupported AVR schema or schema version")
    if not isinstance(payload["assuranceLevel"], str) or not payload["assuranceLevel"]:
        raise ValueError("assuranceLevel must be a non-empty string")
    if not isinstance(payload["issuer"], str) or not ADDRESS.fullmatch(payload["issuer"]):
        raise ValueError("issuer must be a 20-byte EVM address")
    execution = payload["execution"]
    if not isinstance(execution, dict) or set(execution) != {"claimedAt"} or not isinstance(execution["claimedAt"], str):
        raise ValueError("execution must contain only claimedAt")
    try:
        datetime.fromisoformat(execution["claimedAt"].replace("Z", "+00:00"))
    except ValueError as error:
        raise ValueError("claimedAt must be an RFC 3339 timestamp") from error
    if not execution["claimedAt"].endswith("Z"):
        raise ValueError("claimedAt must be UTC and end in Z")
    commitments = payload["commitments"]
    if not isinstance(commitments, dict) or set(commitments) != COMMITMENT_FIELDS:
        raise ValueError("commitments fields do not match the AVR v0.1.0-draft shape")
    if any(not isinstance(value, str) or not BYTES32.fullmatch(value) for value in commitments.values()):
        raise ValueError("Each commitment must be a 32-byte hex value")


def canonicalize(value: object) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def sha256_hex(domain: str, canonical_value: str) -> str:
    return "0x" + hashlib.sha256((domain + canonical_value).encode("utf-8")).hexdigest()


def derive_receipt(receipt: dict) -> dict:
    validate_receipt(receipt)
    payload = {key: value for key, value in receipt.items() if key not in ("expected", "attestation")}
    canonical_receipt = canonicalize(payload)
    canonical_commitments = canonicalize(payload["commitments"])
    return {
        "canonicalReceipt": canonical_receipt,
        "canonicalCommitments": canonical_commitments,
        "receiptId": sha256_hex(RECEIPT_DOMAIN, canonical_receipt),
        "commitmentsRoot": sha256_hex(COMMITMENTS_DOMAIN, canonical_commitments),
    }


def prepare_attestation(receipt: dict) -> dict:
    """Prepare the EIP-191 message for a prototype receipt attestation."""
    receipt_id = derive_receipt(receipt)["receiptId"]
    return {
        "schema": "aichain.avr-attestation",
        "schemaVersion": "0.1.0-draft",
        "scheme": "eip191-personal-sign",
        "receiptId": receipt_id,
        "issuer": receipt.get("issuer"),
        "message": ATTESTATION_MESSAGE_DOMAIN + receipt_id,
    }


def verify_receipt_against_anchor(receipt: dict, anchor: dict) -> dict:
    """Compare a locally derived prototype receipt against a retrieved anchor."""
    derived = derive_receipt(receipt)
    issuer_matches = (
        not anchor.get("issuer")
        or not receipt.get("issuer")
        or anchor["issuer"].lower() == receipt["issuer"].lower()
    )
    result = {
        "receiptId": derived["receiptId"],
        "commitmentsRoot": derived["commitmentsRoot"],
        "receiptIdMatches": not anchor.get("receiptId") or anchor["receiptId"].lower() == derived["receiptId"].lower(),
        "commitmentsRootMatches": anchor["commitmentsRoot"].lower() == derived["commitmentsRoot"].lower(),
        "schemaVersionMatches": anchor["schemaVersion"] == receipt["schemaVersion"],
        "issuerMatches": issuer_matches,
    }
    result["valid"] = all(
        result[key]
        for key in ("receiptIdMatches", "commitmentsRootMatches", "schemaVersionMatches", "issuerMatches")
    )
    return result


def prepare_anchor(receipt: dict) -> dict:
    """Prepare the versioned fields accepted by the prototype anchor contract."""
    derived = derive_receipt(receipt)
    return {
        "receiptId": derived["receiptId"],
        "commitmentsRoot": derived["commitmentsRoot"],
        "schemaVersion": receipt["schemaVersion"],
        "claimedIssuer": receipt.get("issuer"),
    }

File: sdk/typescript/receipt.js
const crypto = require("node:crypto");

const RECEIPT_DOMAIN = "aichain:avr:0.1.0-draft:";
const COMMITMENTS_DOMAIN = "aichain:avr:commitments:0.1.0-draft:";
const ATTESTATION_MESSAGE_DOMAIN = "AIChain AVR v0.1.0-draft receipt: ";
const RECEIPT_FIELDS = new Set(["schema", "schemaVersion", "assuranceLevel", "issuer", "execution", "commitments"]);
const COMMITMENT_FIELDS = new Set(["configuration", "input", "model", "output", "policy", "provider"]);
const BYTES32 = /^0x[0-9a-fA-F]{64}$/;
const ADDRESS = /^0x[0-9a-fA-F]{40}$/;

function hasExactKeys(value, keys) {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    && Object.keys(value).length === keys.size && Object.keys(value).every((key) => keys.has(key));
}

function validateReceipt(receipt) {
  if (receipt === null || typeof receipt !== "object" || Array.isArray(receipt)) throw new Error("Receipt must be an object");
  const { expected, attestation, ...payload } = receipt;
  if (!hasExactKeys(payload, RECEIPT_FIELDS)) throw new Error("Receipt fields do not match the AVR v0.1.0-draft shape");
  if (payload.schema !== "aichain.avr" || payload.schemaVersion !== "0.1.0-draft") throw new Error("Unsupported AVR schema or schema version");
  if (typeof payload.assuranceLevel !== "string" || !payload.assuranceLevel) throw new Error("assuranceLevel must be a non-empty string");
  if (typeof payload.issuer !== "string" || !ADDRESS.test(payload.issuer)) throw new Error("issuer must be a 20-byte EVM address");
  if (!hasExactKeys(payload.execution, new Set(["claimedAt"])) || typeof payload.execution.claimedAt !== "string"
    || !payload.execution.claimedAt.endsWith("Z") || Number.isNaN(Date.parse(payload.execution.claimedAt))) {
    throw new Error("claimedAt must be an RFC 3339 UTC timestamp");
  }
  if (!hasExactKeys(payload.commitments, COMMITMENT_FIELDS) || Object.values(payload.commitments).some((value) => typeof value !== "string" || !BYTES32.test(value))) {
    throw new Error("Each commitment must be a 32-byte hex value");
  }
}

function canonicalize(value) {
  if (Array.isArray(value)) return `[${value.map(canonicalize).join(",")}]`;
  if (value !== null && typeof value === "object") {
    return `{${Object.keys(value).sort().map((key) => `${JSON.stringify(key)}:${canonicalize(value[key])}`).join(",")}}`;
  }
  return JSON.stringify(value);
}

function sha256Hex(domain, canonicalValue) {
  return `0x${crypto.createHash("sha256").update(domain, "utf8").update(canonicalValue, "utf8").digest("hex")}`;
}

function deriveReceipt(receipt) {
  validateReceipt(receipt);
  const { expected, attestation, ...payload } = receipt;
  const canonicalReceipt = canonicalize(payload);
  const canonicalCommitments = canonicalize(payload.commitments);
  return {
    canonicalReceipt,
    canonicalCommitments,
    receiptId: sha256Hex(RECEIPT_DOMAIN, canonicalReceipt),
    commitmentsRoot: sha256Hex(COMMITMENTS_DOMAIN, canonicalCommitments)
  };
}

function prepareAttestation(receipt) {
  const { receiptId } = deriveReceipt(receipt);
  return {
    schema: "aichain.avr-attestation",
    schemaVersion: "0.1.0-draft",
    scheme: "eip191-personal-sign",
    receiptId,
    issuer: receipt.issuer,
    message: `${ATTESTATION_MESSAGE_DOMAIN}${receiptId}`
  };
}

function verifyReceiptAgainstAnchor(receipt, anchor) {
  const derived = deriveReceipt(receipt);
  const issuerMatches = !anchor.issuer || !receipt.issuer || anchor.issuer.toLowerCase() === receipt.issuer.toLowerCase();
  const result = {
    receiptId: derived.receiptId,
    commitmentsRoot: derived.commitmentsRoot,
    receiptIdMatches: !anchor.receiptId || anchor.receiptId.toLowerCase() === derived.receiptId.toLowerCase(),
    commitmentsRootMatches: anchor.commitmentsRoot.toLowerCase() === derived.commitmentsRoot.toLowerCase(),
    schemaVersionMatches: anchor.schemaVersion === receipt.schemaVersion,
    issuerMatches
  };
  return { ...result, valid: result.receiptIdMatches && result.commitmentsRootMatches && result.schemaVersionMatches && result.issuerMatches };
}

function prepareAnchor(receipt) {
  const derived = deriveReceipt(receipt);
  return {
    receiptId: derived.receiptId,
    commitmentsRoot: derived.commitmentsRoot,
    schemaVersion: receipt.schemaVersion,
    claimedIssuer: receipt.issuer
  };
}

module.exports = { canonicalize, validateReceipt, deriveReceipt, prepareAnchor, prepareAttestation, verifyReceiptAgainstAnchor };
